### PR TITLE
[deepspeed doc] install issues + 1-gpu deployment

### PR DESCRIPTION
**Pull request by stas00**
This PR extends the DeepSpeed/FairScale integration documentation to:

* add extensive general troubleshooting for CUDA-extensions (applies to fairscale, deepspeed, apex or any other python pytorch extension with CUDA C++ code) - these are very likely to be encountered by our users - all notes are based on my first hand encounters with these issues - 2 of which I run into yesterday while trying to build fairscale and deepspeed on Sylvain's hardware which he let me use to run the recent benchmarks. so I figured others are likely to have similar issues and neither fairscale nor deepspeed have these documented anywhere.
* adds deployment for 1 gpu DeepSpeed notes 
* reformats sub-headers so that it's easier to link to specific sections

@sgugger

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Great! Thanks for writing down the fixes to the issues you fixed, that's helpful!

File: docs/source/main_classes/trainer.rst [hunk 1]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
 
 This provided support is new and experimental as of this writing.
 
-You will need at least 2 GPUs to benefit from these features.
+Installation Notes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As of this writing, both, FairScale and Deepspeed, require compilation of CUDA C++ code, before they can be used.
+
+While all installation issues should be dealt with through the corresponding github Issues of FairScale and DeepSpeed,
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:121]:
```suggestion
While all installation issues should be dealt with through the corresponding GitHub Issues of FairScale and DeepSpeed,
```

Comment (sgugger) [reply docs/source/main_classes/trainer.rst:121]: + links to issues of FairScale and DeepSpeed (under their respective names) here?

File: docs/source/main_classes/trainer.rst [hunk 2]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
...
+there are a few common issues that one may encounter while building any pytorch extension that needs to build CUDA
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:122]:
```suggestion
there are a few common issues that one may encounter while building any PyTorch extension that needs to build CUDA
```

File: docs/source/main_classes/trainer.rst [hunk 3]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
...
+    pip install deepspeed
+
+please, read the following notes first.
+
+In these notes we give examples for what to do when ``pytorch`` has been built with CUDA ``10.2``. If your situation is
+different remember to adjust the version number to the one you are after.
+
+**Possible problem #1:**
+
+While, pytorch comes with its own CUDA toolkit, to build these two projects you must have an identical version of CUDA
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:139]:
```suggestion
While, PyTorch comes with its own CUDA toolkit, to build these two projects you must have an identical version of CUDA
```
(Everywhere in text pytorch -> PyTorch, the \`\`pytorch\`\` in backsticks can stay the same, will stop highlighting)

File: docs/source/main_classes/trainer.rst [hunk 4]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
...
+installed system-wide.
+
+For example, if you installed ``pytorch`` with ``cudatoolkit==10.2`` in the python environment, you also need to have
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:142]:
```suggestion
For example, if you installed ``pytorch`` with ``cudatoolkit==10.2`` in the Python environment, you also need to have
```
Everywhere python -> Python while we are at it :-)

File: docs/source/main_classes/trainer.rst [hunk 5]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
...
+
+Here, we are making a symlink to ``gcc-7`` from ``/usr/local/cuda-10.2/bin/gcc`` and since
+``/usr/local/cuda-10.2/bin/`` should be in the ``PATH`` environment variable (see the previous problem's solution), it
+should find ``gcc-7`` (and ``g++7``) and then the build will succeed.
+
+As always make sure to edit the paths in the example to match your situation.
+
+**If still unsuccessful:**
+
+If after addressing these you still encounter build issues, please, proceed with the github Issue of `FairScale
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:223]:
```suggestion
If after addressing these you still encounter build issues, please, proceed with the GitHub Issue of `FairScale
```

File: docs/source/main_classes/trainer.rst [hunk 6]
```diff
@@ -124,6 +234,8 @@ provides support for the following features from `the ZeRO paper <https://arxiv.
 1. Optimizer State Sharding
 2. Gradient Sharding
 
+You will need at least 2 GPUs to use this feature.
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:237]:
```suggestion
You will need at least two GPUs to use this feature.
```

File: docs/source/main_classes/trainer.rst [hunk 7]
```diff
@@ -164,66 +276,126 @@ Notes:
...
+Install the library via pypi:
 
-   .. code-block:: bash
+.. code-block:: bash
 
-       pip install deepspeed
+    pip install deepspeed
 
-   or find more details on `the DeepSpeed's github page <https://github.com/microsoft/deepspeed#installation>`__.
+or find more details on `the DeepSpeed's github page <https://github.com/microsoft/deepspeed#installation>`__.
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:295]:
```suggestion
or find more details on `the DeepSpeed's GitHub page <https://github.com/microsoft/deepspeed#installation>`__.
```
Everywhere github -> GitHub ;-)

File: docs/source/main_classes/trainer.rst [hunk 8]
```diff
@@ -164,66 +276,126 @@ Notes:
...
+Note that in the DeepSpeed documentation you are likely to see ``--deepspeed --deepspeed_config ds_config.json`` - i.e.
+two DeepSpeed-related arguments, but for the sake of simplicity, and since there are already so many arguments to deal
+with, we combined the two into a single argument.
+
+For some practical usage examples, please, see this `post
+<https://github.com/huggingface/transformers/issues/8771#issuecomment-759248400>`__.
+
+
+
+Deployment with 1 GPU
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:345]:
```suggestion
Deployment with one GPU
```

File: docs/source/main_classes/trainer.rst [hunk 9]
```diff
@@ -164,66 +276,126 @@ Notes:
...
+=======================================================================================================================
+
+To deploy DeepSpeed with 1 GPUs adjust the :class:`~transformers.Trainer` command line arguments as following:
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:348]:
```suggestion
To deploy DeepSpeed with one GPU adjust the :class:`~transformers.Trainer` command line arguments as following:
```

File: docs/source/main_classes/trainer.rst [hunk 10]
```diff
@@ -164,66 +276,126 @@ Notes:
...
+    --do_train --n_train 500 --num_train_epochs 1 \
+    --per_device_train_batch_size 1  --freeze_embeds \
+    --src_lang en_XX --tgt_lang ro_RO --task translation
+
+This is almost the same as with multiple-GPUs, but here we tell DeepSpeed explicitly to use just one GPU. By default,
+DeepSpeed deploys all GPUs it can see. If you have only 1 GPU to start with, then you don't need this argument. The
+following `documentation <https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node>`__ discusses the
+launcher options.
+
+Why would you want to use DeepSpeed with just 1 GPU?
```
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:365]:
```suggestion
Why would you want to use DeepSpeed with just one GPU?
```

**Review by sgugger — APPROVED** [review]
Thanks a lot for expanding the doc! Just left some nits (mostly about the correct case of some names).

File: docs/source/main_classes/trainer.rst [hunk 11]
```diff
@@ -113,7 +113,117 @@ Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, O
...
+While, pytorch comes with its own CUDA toolkit, to build these two projects you must have an identical version of CUDA
+installed system-wide.
+
+For example, if you installed ``pytorch`` with ``cudatoolkit==10.2`` in the python environment, you also need to have
+CUDA ``10.2`` installed system-wide.
+
+The exact location may vary from system to system, but ``/usr/local/cuda-10.2`` is the most common location on many
+Unix systems. If you don't have it installed, install it first. You will find the instructions by using your favorite
+search engine. For example, if you're on Ubuntu you may want to search for: `ubuntu cuda 10.2 install
+<https://www.google.com/search?q=ubuntu+cuda+10.2+install>`__.
```
[the remark this answers is not in the document]
Comment (stas00) [reply docs/source/main_classes/trainer.rst:148]: great tip - added!

**Comment by stas00** [comment]
Thank you for your awesome suggestions and tweaks - all done.

**Pull request merged**